traitsui/qt4/key_binding_editor.py

```diff
@@ -57,47 +58,34 @@ def init(self, parent):
         """
         self.control = KeyBindingCtrl(self)
 
-    def update_object(self, event):
-        """Handles the user entering input data in the edit control."""
-        try:
-            self.value = value = key_event_to_name(event)
-            self._binding.text = value
-        except:
-            pass
-
     def update_editor(self):
         """Updates the editor when the object trait changes externally to the
         editor.
         """
         self.control.setText(self.value)
 
-    def update_focus(self, has_focus):
-        """Updates the current focus setting of the control."""
-        if has_focus:
-            self._binding.border_size = 1
-            self.object.owner.focus_owner = self._binding
-
     def _key_changed(self, event):
         """Handles a keyboard event."""
         binding = self.object
         key_name = key_event_to_name(event)
-        cur_binding = binding.owner.key_binding_for(binding, key_name)
+        cur_binding = self.ui.parent.handler.key_binding_for(binding, key_name)
         if cur_binding is not None:
-            if (
-                QtGui.QMessageBox.question(
-                    self.control,
-                    "Duplicate Key Definition",
-                    "'%s' has already been assigned to '%s'.\n"
+            result = confirm(
+                parent=self.control,
+                message=(
+                    f"{key_name!r} has already been assigned to "
+                    f"'{cur_binding.description}'.\n"
                     "Do you wish to continue?"
-                    % (key_name, cur_binding.description),
-                    QtGui.QMessageBox.StandardButton.Yes | QtGui.QMessageBox.StandardButton.No,
-                    QtGui.QMessageBox.StandardButton.No,
-                )
-                != QtGui.QMessageBox.StandardButton.Yes
-            ):
+                ),
+                title="Duplicate Key Definition",
+            )
+            if result != YES:
                 return
 
         self.value = key_name
+        # Need to manually update editor because loopback protection doesn't
```

Comment:
what does loopback protection mean here?

Reply:
I think it means trait change listeners are disabled when updating the `Editor.value` trait - IIRC it's done to prevent recursion of the listeners or something of that sort.

Reply:
I think my comment here is a bit garbled.  I'll update.

As Rahul mentioned, you frequently run into trouble in GUI code with infinite event loops where change in UI -> change in model -> change in UI -> ... etc.  So TraitsUI has code to prevent this sort of loopback.  When you set the `value`, it assumes that it is coming from the toolkit, and so the toolkit objects are in the desired state, and so it generally blocks pushing back changes to the toolkit object from this change.

However in this case, the content of the toolkit object isn't matching the value, so we need to manually update that as well.

This isn't needed on wx, since it overrides drawing completely and gets the context from the editor value.